tokenbucket.go

```diff
@@ -465,3 +469,132 @@ func (t *TokenBucketRedis) SetState(ctx context.Context, state TokenBucketState)
+// PartitionKey is the key used to store all the this implementation in DynamoDB.
+//
+// TableProps describe the table that this backend should work with. This backend requires the following on the table:
+// * TTL
+//
+// TTL is the TTL of the stored item.
+//
+// If raceCheck is true and the item in DynamoDB are modified in between State() and SetState() calls then
+// ErrRaceCondition is returned.
+func NewTokenBucketDynamoDB(client *dynamodb.Client, partitionKey string, tableProps DynamoDBTableProperties, ttl time.Duration, raceCheck bool) *TokenBucketDynamoDB {
+	keys := map[string]types.AttributeValue{
+		tableProps.PartitionKeyName: &types.AttributeValueMemberS{Value: partitionKey},
+	}
+
+	if tableProps.SortKeyUsed {
+		keys[tableProps.SortKeyName] = &types.AttributeValueMemberS{Value: partitionKey}
+	}
+
+	return &TokenBucketDynamoDB{
+		client:       client,
+		partitionKey: partitionKey,
+		tableProps:   tableProps,
+		ttl:          ttl,
+		raceCheck:    raceCheck,
+		keys:         keys,
+	}
+}
+
+// State gets the bucket's state from DynamoDB.
+func (t *TokenBucketDynamoDB) State(ctx context.Context) (TokenBucketState, error) {
+	resp, err := dynamoDBGetItem(ctx, t.client, t.getGetItemInput())
+
+	if err != nil {
+		return TokenBucketState{}, err
+	}
+
+	return t.loadStateFromDynamoDB(resp)
+}
+
+// SetState updates the state in DynamoDB.
+func (t *TokenBucketDynamoDB) SetState(ctx context.Context, state TokenBucketState) error {
+	input := t.getPutItemInputFromState(state)
+
+	var err error
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		_, err = dynamoDBputItem(ctx, t.client, input)
+	}()
+
+	select {
+	case <-done:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	return err
+}
+
+const dynamoDBBucketAvailableKey = "Available"
+
+func (t *TokenBucketDynamoDB) getGetItemInput() *dynamodb.GetItemInput {
+	return &dynamodb.GetItemInput{
+		TableName: &t.tableProps.TableName,
+		Key:       t.keys,
+	}
+}
+
+func (t *TokenBucketDynamoDB) getPutItemInputFromState(state TokenBucketState) *dynamodb.PutItemInput {
+	item := map[string]types.AttributeValue{}
+	for k, v := range t.keys {
+		item[k] = v
+	}
+
+	item[dynamoDBBucketLastKey] = &types.AttributeValueMemberN{Value: strconv.Itoa(int(state.Last))}
+	item[dynamoDBBucketVersionKey] = &types.AttributeValueMemberN{Value: strconv.Itoa(int(t.latestVersion) + 1)}
+	item[t.tableProps.TTLFieldName] = &types.AttributeValueMemberN{Value: strconv.Itoa(int(time.Now().Add(t.ttl).Unix()))}
+	item[dynamoDBBucketAvailableKey] = &types.AttributeValueMemberN{Value: strconv.Itoa(int(state.Available))}
+
+	input := &dynamodb.PutItemInput{
+		TableName: &t.tableProps.TableName,
+		Item:      item,
+	}
+
+	if t.raceCheck && t.latestVersion > 0 {
+		input.ConditionExpression = aws.String(dynamodbBucketRaceConditionExpression)
+		input.ExpressionAttributeValues = map[string]types.AttributeValue{
+			":version": &types.AttributeValueMemberN{Value: strconv.Itoa(int(t.latestVersion))},
```

Comment:
Redundant type conversion: 

```suggestion
			":version": &types.AttributeValueMemberN{Value: strconv.Itoa(t.latestVersion)},
```

Reply:
removed